docs/enhancements/agent-late-binding.md

```diff
@@ -0,0 +1,126 @@
+1. The events for an Agent which is not associated with a cluster, for example a history of when it registered and which clusters it was associated with (should be visible to the infrastructure admin).  These events should be scoped to an InfraEnv.
+
+### User Stories
+
+#### Story 1
+
+As an Infrastructure Admin, I want to create a Discovery ISO and use it to boot hosts for use by Cluster Creators to install OpenShift on.
+
+#### Story 2
+
+As an Infrastructure Admin, I want to add Discovery ISOs to hosts as a boot option, and then ship the hosts to remote locations where they will be booted at a later time for use by Cluster Creators to install OpenShift on.
+
+#### Story 3
+
+As a Cluster Creator, I want to view a collection of available Agents and use them for OpenShift cluster creation or expansion.
+
+#### Story 4
+
+As a Cluster Creator, I want to reassign an Agent from one OpenShift cluster to another (not during installation).
+
+### Implementation Details/Notes/Constraints
+
+It would be beneficial to move the discovery image management to a separate service at some point, both to allow it to be scaled independently and to reduce the very large scope of the existing service.  The late binding work should be done with this separation in mind, even if the actual separation is an orthogonal task.
+
+### Risks and Mitigations
+
+## Design Details
+
+- Currently, a cluster ID is baked into the ISO, and the agent relays that to the service upon registration.  With late binding, the InfraEnv ID should replace the cluster ID in this context.  The controller will also use this information to create the Agent CRs in the same namespaces as their associated InfraEnvs.
+
+- Internally, in the SQL DB, hosts currently have two primary keys - the host ID and the cluster ID.  Now the primary keys will be host ID and InfraEnv ID.
+
+- Currently the agent writes the image from the ISO to disk, which means all Agents that form a cluster need to be running with the same version.  Instead, the agent should download the correct version for the cluster being formed if it doesn't match what's in the ISO.
+
+- When moving a host from a source to dest cluster the service needs to make sure that the source cluster is not installing (can be before or after) and that the dest cluster is before installation.  If the source cluster is before installation, the host will automatically go to disconnected state, which is the correct behavior.  However, the service should update its state manually so it doesn't take 3+ minutes (e.g., by resetting the timestamp of when it last checked in).  If the host doesn't exist in the dest cluster a new host will be created (register_host).  If it does exist, we reset its state and properties but not its events.
+
+- There should be one pull secret for the ISO to pull the necessary images (Infrastructure Admin's'), and another pull secret in the cluster definition for the installed cluster to use (Cluster Creator's).
```

Comment:
This is a great distinction. I never understood why the pull secret was needed for both the image and the cluster (I think they're typically the same today), but this makes a lot of sense.